docker/daemon/daemon_dest.go

```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
 		defer os.Remove(streamCopy.Name())
 		defer streamCopy.Close()
 
-		size, err := io.Copy(streamCopy, stream)
+		// If we need to compute a digest.
+		digester := digest.Canonical.New()
+		tee := io.TeeReader(stream, digester.Hash())
+
+		size, err := io.Copy(streamCopy, tee)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
+			inputInfo.Digest = digester.Digest()
```

Comment:
Actually the whole handling of `Digest` values in this function, especially immediately below, seems to be broken (should be a separate PR) and only happens to work through a random coincidence:

The returned digest must match the file name within the tarball (because `PutManifest` uses the layer digests to compute `layerPaths`. But, right now, we create the file with `inputInfo.Digest` (using, in principle, any algorithm), but we always return a freshly computed `digest.Canonical` value; the two do not generally match.

Basically, if we are provided a digest, we should either use it as is, _and return_, or we need to completely ignore it and compute from scratch (but the rest of the code generally expects us not to do that, for signatures etc.).

As an extension, this also works only because `inputInfo.Digest` is computed using `digest.Canonical`, just like the the computation below; that is, I expect, by design, but the code does not make that obvious (and again, computing twice…)